driver-core/src/main/com/mongodb/internal/connection/DefaultServer.java

```diff
@@ -67,41 +85,67 @@
         serverListener.serverOpening(new ServerOpeningEvent(this.serverId));
 
         this.serverMonitor = serverMonitor;
+        operationCount = trackOperationCount ? new AtomicInteger() : null;
     }
 
     @Override
     public Connection getConnection() {
         isTrue("open", !isClosed());
         SdamIssue.Context exceptionContext = sdam.context();
         try {
-            return connectionFactory.create(connectionPool.get(), new DefaultServerProtocolExecutor(), clusterConnectionMode);
+            operationBegin();
+            return OperationCountTrackingConnection.decorate(this,
+                    connectionFactory.create(connectionPool.get(), new DefaultServerProtocolExecutor(), clusterConnectionMode));
         } catch (MongoException e) {
+            operationEnd();
             sdam.handleExceptionBeforeHandshake(SdamIssue.specific(e, exceptionContext));
             throw e;
+        } catch (RuntimeException e) {
+            operationEnd();
```

Comment:
I will change the code, but I want to say that I am not handling `Error`s not because I rely on the user closing the client or terminating the process, but because our (and highly likely user's) code already fails to survive many trivial `Error`s. As an example, if `connectionPool.getAsync` throws an OOM while [creating a new `ConnectionCheckOutStartedEvent`](https://github.com/stIncMale/mongo-java-driver/blob/JAVA-3926/driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java#L181), then we will not call `operationEnd` in [`DefaultServer.getConnectionAsync`](https://github.com/stIncMale/mongo-java-driver/blob/JAVA-3926/driver-core/src/main/com/mongodb/internal/connection/DefaultServer.java#L114).

In order to avoid problems like the one described above, we would need to handle abrupt completions of all asynchronous methods (i.e. not only handle errors in callbacks, but also directly in the code that calls async methods), which we do not do, and frankly, doing so would make the async code even less maintainable. But this was just a trivial example, and one may think of a more convoluted way to end up in a corrupt state after an `Error`.

So as always, I am saying that the ability to handle `Errors` gives a false impression that the code can be easily written in such a way that it survives an `Error`, while, for example, most of real asynchronous code fails to do so.

There can be argument "let's handle what we can handle easily, thus reducing the probability of having a corrupt state after an `Error`", which, if followed (and virtually everyone follows this approach), leads to every application at best logging `Error`s and continuing working, expecting that everything is fine. Which is very similar to relying on a specific behavior in a program written in a language with UB, when the specification states that the only guarantee is that the behavior is undefined - it works until it fails terribly.

Reply:
Done.